machine_learning/a_star_search.cpp

```diff
@@ -0,0 +1,659 @@
+/**
+ * @brief
+ * [A* search algorithm](https://en.wikipedia.org/wiki/A*_search_algorithm)
+ * @details
+ * A* is an informed search algorithm, or a best-first search, meaning that it
+ * is formulated in terms of weighted graphs: starting from a specific starting
+ * node of a graph (initial state), it aims to find a path to the given goal
+ * node having the smallest cost (least distance travelled, shortest time,
+ * etc.). It evaluates by maintaining a tree of paths originating at the start
+ * node and extending those paths one edge at a time until it reaches the final
+ * state.
+ * The weighted edges (or cost) is evaluated on two factors, G score
+ * (cost required from starting node or initial state to current state) and H
+ * score (cost required from current state to final state). The F(state), then
+ * is evaluated as:
+ * F(state) = G(state) + H(state).
+ *
+ * To solve the given search with shortest cost or path possible  is to inspect
+ * values having minimum F(state).
+ * @author [Ashish Daulatabad](https://github.com/AshishYUO)
+ */
+#include <algorithm>   /// for `std::reverse` function
+#include <array>       /// for `EightPuzzle` board
+#include <cassert>     /// for `assert`
+#include <functional>  /// for `std::function` STL
+#include <iostream>    /// for IO operations
+#include <map>         /// for `std::map` STL
+#include <set>         /// for `std::set` STL
+#include <vector>      /// for `std::vector` STL
+/**
+ * @namespace machine_learning
+ * @brief Machine learning algorithms
+ */
+namespace machine_learning {
+/**
+ * @namespace aystar_search
+ * @brief Functions for [A*
+ * Search](https://en.wikipedia.org/wiki/A*_search_algorithm).
+ */
+namespace aystar_search {
+/**
+ * @class EightPuzzle
+ * @details
+ * A well known 3 x 3 puzzle of the form
+ * `
+ * 1   2   3
+ * 4   5   6
+ * 7   8   0
+ * `
+ * where `0` represents an empty space in the puzzle
+ * Given any random state, the goal is to achieve the above configuration
+ * (or any other configuration if possible)
+ * @tparam N size of the square Puzzle, default is set to 3 (since it is
+ * EightPuzzle)
+ */
+template <size_t N = 3>
+class EightPuzzle {
+    // N x N array to store the current state of the Puzzle.
+    std::array<std::array<int, N>, N> board;
```

Comment:
Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂